boilerplate/app/utils/useAppTheme.ts

```diff
@@ -0,0 +1,109 @@
+import { DarkTheme, DefaultTheme, useTheme as useNavTheme } from "@react-navigation/native"
+import {
+  type Theme,
+  type ThemeContexts,
+  type ThemedStyle,
+  type ThemedStyleArray,
+  lightTheme,
+  darkTheme,
+} from "app/theme"
+
+type ThemeContextType = {
+  themeScheme: ThemeContexts
+  setThemeContextOverride: (newTheme: ThemeContexts) => void
+}
+
+// create a React context and provider for the current theme
+export const ThemeContext = createContext<ThemeContextType>({
+  themeScheme: undefined, // default to the system theme
+  setThemeContextOverride: (_newTheme: ThemeContexts) => {
+    console.error("Tried to call setThemeContextOverride before the ThemeProvider was initialized")
+  },
+})
+
+export const useThemeProvider = (initialTheme: ThemeContexts = undefined) => {
+  const colorScheme = useColorScheme()
+  const [overrideTheme, setTheme] = useState<ThemeContexts>(initialTheme)
+
+  const setThemeContextOverride = useCallback((newTheme: ThemeContexts) => {
+    setTheme(newTheme)
+  }, [])
+
+  const themeScheme = overrideTheme || colorScheme || "light"
+  const navigationTheme = themeScheme === "dark" ? DarkTheme : DefaultTheme
+
+  return {
+    themeScheme,
+    navigationTheme,
+    setThemeContextOverride,
+    ThemeProvider: ThemeContext.Provider,
+  }
+}
+
+interface UseAppThemeValue {
+  // The theme object from react-navigation
+  navTheme: typeof DefaultTheme
+  // A function to set the theme context override (for switching modes)
+  setThemeContextOverride: (newTheme: ThemeContexts) => void
+  // The current theme object
+  theme: Theme
+  // The current theme context "light" | "dark"
+  themeContext: ThemeContexts
+  // A function to apply the theme to a style object.
+  // See examples in the components directory or read the docs here:
+  // https://docs.infinite.red/ignite-cli/boilerplate/app/utils/
+  themed: <T>(styleOrStyleFn: ThemedStyle<T> | StyleProp<T> | ThemedStyleArray<T>) => T
+}
+
+/**
+ * Custom hook that provides the app theme and utility functions for theming.
+ *
+ * @returns {UseAppThemeReturn} An object containing various theming values and utilities.
+ * @throws {Error} If used outside of a ThemeProvider.
+ */
+export const useAppTheme = (): UseAppThemeValue => {
+  const navTheme = useNavTheme()
+  const context = useContext(ThemeContext)
+  if (!context) {
+    throw new Error("useTheme must be used within a ThemeProvider")
+  }
+
+  const { themeScheme: overrideTheme, setThemeContextOverride } = context
+
+  const themeContext: ThemeContexts = useMemo(
+    () => overrideTheme || (navTheme.dark ? "dark" : "light"),
+    [overrideTheme, navTheme],
+  )
+
+  const themeVariant: Theme = useMemo(
+    () => (themeContext === "dark" ? darkTheme : lightTheme),
+    [themeContext],
+  )
+
+  const themed = useCallback(
+    <T>(styleOrStyleFn: ThemedStyle<T> | StyleProp<T> | ThemedStyleArray<T>) => {
+      const flatStyles = [styleOrStyleFn].flat(3) as (ThemedStyle<T> | StyleProp<T>)[]
```

Comment:
I just checked if this cast was necessary, I removed it and am not seeing any issues, okay to remove @markrickert ?

Reply:
Nice find! Lets get rid of it